components/eamxx/src/share/field/field_impl.hpp

```diff
@@ -0,0 +1,587 @@
+  //       reshaped view first, based on the field rank.
+
+  const auto& layout = get_header().get_identifier().get_layout();
+  const auto  rank   = layout.rank();
+  switch (rank) {
+    case 1:
+      {
+        auto v = get_view<ST*,HD>();
+        Kokkos::deep_copy(v,value);
+      }
+      break;
+    case 2:
+      {
+        auto v = get_view<ST**,HD>();
+        Kokkos::deep_copy(v,value);
+      }
+      break;
+    case 3:
+      {
+        auto v = get_view<ST***,HD>();
+        Kokkos::deep_copy(v,value);
+      }
+      break;
+    case 4:
+      {
+        auto v = get_view<ST****,HD>();
+        Kokkos::deep_copy(v,value);
+      }
+      break;
+    case 5:
+      {
+        auto v = get_view<ST*****,HD>();
+        Kokkos::deep_copy(v,value);
+      }
+      break;
+    case 6:
+      {
+        auto v = get_view<ST******,HD>();
+        Kokkos::deep_copy(v,value);
+      }
+      break;
+    default:
+      EKAT_ERROR_MSG ("Error! Unsupported field rank in 'deep_copy'.\n");
+  }
+}
+
+template<HostOrDevice HD, typename ST>
+void Field::
+update (const Field& x, const ST alpha, const ST beta)
+{
+  // Check x/y are allocated
+  EKAT_REQUIRE_MSG (is_allocated(),
+      "Error! Cannot update field, since it is not allocated.\n"
+      " - field name: " + name() + "\n");
+  EKAT_REQUIRE_MSG (x.is_allocated(),
+      "Error! Cannot update field, since source field is not allocated.\n"
+      " - field name: " + x.name() + "\n");
+
+  // Check y is writable
+  EKAT_REQUIRE_MSG (not is_read_only(),
+      "Error! Cannot update field, as it is read-only.\n"
+      " - field name: " + name() + "\n");
+
+  // Check compatibility between fields data type
+  const auto dt_x = x.data_type();
+  const auto dt_y = data_type();
+
+  // We could be more general, and do
+  //   EKAT_REQUIRE_MSG (not is_narrowing_conversion(dt_x,dt_y),"");
+  // but then we'd need N*N template instantiations of the impl function
+  EKAT_REQUIRE_MSG (dt_x==dt_y,
+      "Error! X and Y field must have the same data type.\n"
+      " - x name: " + x.name() + "\n"
+      " - y name: " + name() + "\n"
+      " - x data type: " + e2str(dt_x) + "\n"
+      " - y data type: " + e2str(dt_y) + "\n");
+
+  // If user used default for both alpha and beta, then the ST=char was used,
+  // even though that's not correct; hence, set data type of coeffs to that of x/y.
+  const auto dt_st = get_data_type<ST>();
+  EKAT_REQUIRE_MSG (not is_narrowing_conversion(dt_st,dt_x),
+      "Error! Coefficients alpha/beta may be narrowed when converted to x/y data type.\n"
+      " - x/y data type  : " + e2str(dt_x) + "\n"
+      " - coeff data type: " + e2str(dt_st) + "\n");
+
+  if (dt_x==DataType::IntType) {
+    return update_impl<HD,int>(x,alpha,beta);
+  } else if (dt_x==DataType::FloatType) {
+    return update_impl<HD,float>(x,alpha,beta);
+  } else if (dt_x==DataType::DoubleType) {
+    return update_impl<HD,double>(x,alpha,beta);
+  }
```

Comment:
Currently a Field requires one of these 3 data types, but could this change in the future? If a new DT is added without updating this function, `update()` would do nothing (but silently) since `update_impl()` just would not be called. Should we `else { Error("invalid data type") }` here? Then if a new DT is added, anyone trying to use this function would at least be tipped off that something is wrong.

Reply:
Good catch!